src/main/java/hudson/remoting/RemoteInvocationHandler.java

```diff
@@ -233,16 +292,217 @@ public int hashCode() {
+            this.origin = referent.origin;
+            this.channel = referent.channel;
+        }
+
+        /**
+         * Sends the {@link UnexportCommand} for the specified {@link #oid} if the {@link Channel} is still open.
+         * @throws IOException if the {@link UnexportCommand} could not be sent.
+         */
+        private void cleanup() throws IOException {
+            if (this.channel == null) {
+                return;
+            }
+            Channel channel = this.channel.channel();
+            if (channel != null && !channel.isClosingOrClosed()) {
+                try {
+                    channel.send(new UnexportCommand(oid, origin));
+                } finally {
+                    // clear out references to simplify GC
+                    this.origin = null;
+                    this.channel = null;
+                }
+            }
         }
-        super.finalize();
     }
 
+    /**
+     * Manages the cleanup of {@link RemoteInvocationHandler} instances that need to be auto unexported.
+     * @since FIXME after merge
+     */
+    private static class Unexporter implements Runnable {
+
+        /**
+         * Our executor service, we use at most one thread for all {@link Channel} instances in the current classloader.
+         */
+        private final ExecutorService svc = new AtmostOneThreadExecutor(
+                new NamingThreadFactory(new DaemonThreadFactory(), RemoteInvocationHandler.class.getSimpleName())
+        );
+        /**
+         * Flag to track that {@link #UNEXPORTER} has been queued for execution.
+         */
+        private final AtomicBoolean inQueue = new AtomicBoolean(false);
+        /**
+         * Flag to track that {@link #UNEXPORTER} is running.
+         */
+        private final AtomicBoolean isAlive = new AtomicBoolean(false);
+        /**
+         * Our {@link ReferenceQueue} for picking up references that have been collected by the garbage collector
+         * and need the corresponding {@link UnexportCommand} sent.
+         */
+        private final ReferenceQueue<? super RemoteInvocationHandler> queue = new ReferenceQueue<RemoteInvocationHandler>();
+        /**
+         * The "live" {@link PhantomReferenceImpl} instances for each active {@link Channel}.
+         */
+        private final ConcurrentMap<Channel.Ref,List<PhantomReferenceImpl>> referenceLists 
+                = new ConcurrentHashMap<Channel.Ref, List<PhantomReferenceImpl>>();
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public void run() {
+            if (!isAlive.compareAndSet(false, true)) {
+                inQueue.set(false); // we have started execution so clear the flag, must happen after check of running
+                return;
+            }
+            inQueue.set(false); // we have started execution and running is true, so queued can be reset
+            try {
+                while (!referenceLists.isEmpty()) {
```

Comment:
395-426 gets one reference from the queue and cleans it up. 429-439 scans for any channels that have closed and removes their objects from even the phantom reference queue thus making it super easy for the GC to collect them... though having looked at that code in the cold light of day I now want to wrap the second one in a periodic run basis rather than every loop in case there are a large number of references cleared when there are many channels open